p4-fusion/main.cc

```diff
@@ -63,9 +64,77 @@ int Main(int argc, char** argv)
 		return 1;
 	}
 
-	// Set the signal here because it gets reset after P4API library is initialized.
-	std::signal(SIGINT, SignalHandler);
-	std::signal(SIGTERM, SignalHandler);
+	// Block signals from being handled by the main thread, and all future threads.
+	//
+	// - SIGINT, SIGTERM, SIGHUP: thread pool will be shutdown and std::exit will be called
+	// - SIGUSR1: only sent by main() to tell the signal handler thread to exit
+	sigset_t blockedSignals;
+	sigemptyset(&blockedSignals);
+	for (auto sig : { SIGINT, SIGTERM, SIGHUP, SIGUSR1 })
+	{
+		sigaddset(&blockedSignals, sig);
+	}
+
+	int rc = pthread_sigmask(SIG_BLOCK, &blockedSignals, nullptr);
+	if (rc != 0)
+	{
+		ERR("(signal handler) failed to block signals: (" << errno << ") " << strerror(errno));
+		return 1;
+	}
+
+	// Create the thread pool
+	int networkThreads = arguments.GetNetworkThreads();
+	PRINT("Network Threads: " << networkThreads);
+	PRINT("Creating " << networkThreads << " network threads");
+	ThreadPool pool(networkThreads);
+	SUCCESS("Created " << pool.GetThreadCount() << " threads in thread pool");
+
+	// Spawn a thread to handle signals.
+	// The thread will block and wait for signals to arrive and then shutdown the thread pool, unless it receives SIGUSR1,
+	// in which case it will just exit (since main() is handling the shutdown).
+	//
+	// Using a separate thread for purely signal handling allows us to use non-reentrant functions
+	// (such as std::cout, condition variables, etc.) in the signal handler.
+	auto signalHandlingThread = std::thread(
+	    [&blockedSignals, &pool]()
+	    {
+		    auto shutdown = [&pool]()
```

Comment:
Suggestion: Factor out this `shutdown` function to be a top-level function which takes an exit code and a reference to the pool, have it call `std::exit()`. Ideally, all shutdown codepaths should be directed to the same shutdown function, but there is slightly different logic inside this `shutdown()` lambda compared to the end of `main` right now.

Reply:
[`d8159cb` (#9)](https://github.com/sourcegraph/p4-fusion/pull/9/commits/d8159cbfba53d64f535cc9fbc3215165c0e38821)

Do you mean this? I'm happy to make the change, but is there a stylistic / convention reason you're looking for this v. a lambda?

Reply:
> Do you mean this?

Yes, but also to replace this at the end of `main()`

```
	pool.ShutDown();
	// Wait for exception handler to finish.
	t.join();
	if (!P4API::ShutdownLibraries())
	{
		return 1;
	}
	// Finalize tracing.
	mtr_flush();
	mtr_shutdown();
```

With a call to the `shutdown()` function.